### PR TITLE
Refactor Settings

### PR DESCRIPTION
**Pull request by alejoberardino**
The current implementation of custom settings caused a lot of issues with generics. The problem was mainly around the fact that the type for the resolved settings (that is, the key:value pair once processed) was also itself a generic parameter, and it really messed up with polymorphism. Things like 
```typescript
const adapter = new Adapter({
	name: 'TEST',
	endpoints: [customEndpoint] // Endpoint is of type CustomEndpoint, with typeof customSettings
	customSettings: customSettings
})
```
Would fail in the assignment of the endpoint to the array of endpoints for the adapter parameters, due to a clash in settings. Another thing was a bit annoying, in that in order to define the settings in a way that added type safety one needed to make use of the `satisfies` keyword. There is nothing necessarily wrong with that, but it's a newer feature and developers might not be very familiar with it. 

This PR refactors the way the adapter config is defined, providing a class based approach that allows for cleaner generics and a type safe way to define them without asking a lot of the developer.

### CODE REVIEW

File: src/adapter/types.ts [hunk 1]
```diff
@@ -99,17 +99,14 @@ export interface AdapterParams<CustomSettings extends SettingsMap> {
...
-  customSettings?: SettingsMap
-
   /** Configuration relevant to outbound (EA --\> DP) communication rate limiting */
   rateLimiting?: AdapterRateLimitingConfig
 
   /** Bootstrap function that will run when initializing the adapter */
-  bootstrap?: (adapter: Adapter<CustomSettings>) => Promise<void>
+  bootstrap?: (adapter: Adapter<T>) => Promise<void>
+
+  /** TODO: complete */
```
Comment (austinborn) [inline src/adapter/types.ts:108]: TODO - can you either finish this or add a ticket for it (if there isn't one already)?

File: src/config/index.ts [hunk 2]
```diff
@@ -519,11 +506,98 @@ export type AdapterConfigFromSettings<T extends SettingsMap> = {
...
+export type EmptySettingsDefinitionMap = Record<string, never>
+export type SettingsDefinitionMap = Record<string, SettingDefinition>
 export type ValidationErrorMessage = string | undefined
+
+/**
+ * TODO: Improve docs
+ * This class will hold the processed config type, and the basic settings
+ * The idea is that you can no longer use a straight object, but have to build a config
+ * Then in the generics we can simply pass the type of this, and it will hopefully allow for simpler generics
+ * name is placeholder
```
Comment (austinborn) [inline src/config/index.ts:523]: Name still placeholder?

File: src/config/index.ts [hunk 3]
```diff
@@ -519,11 +506,98 @@ export type AdapterConfigFromSettings<T extends SettingsMap> = {
...
+export class AdapterConfig<T extends SettingsDefinitionMap = SettingsDefinitionMap> {
+  settings!: AdapterSettings<T>
+
+  constructor(
+    private settingsDefinition: T,
+    private options?: {
+      /** Map of overrides to the default config values for an Adapter */
+      envDefaultOverrides?: Partial<BaseAdapterSettings>
+
+      /** TODO: complete */
```
Comment (austinborn) [inline src/config/index.ts:534]: Another TODO

File: src/config/index.ts [hunk 4]
```diff
@@ -519,11 +506,98 @@ export type AdapterConfigFromSettings<T extends SettingsMap> = {
...
+          setting.sensitive &&
+          (this.settings as Record<string, ValidSettingValue>)[name],
+      )
+      .map(([name]) => ({
+        key: name,
+        // Escaping potential special characters in values before creating regex
+        value: new RegExp(
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          ((this.settings as AdapterSettings)[name]! as string).replace(
+            /[-[\]{}()*+?.,\\^$|#\s]/g,
```
Comment (austinborn) [inline src/config/index.ts:589]: Did you come up with this or pull it from somewhere?

**Review by austinborn — APPROVED** [review]
Looks fine, just a couple open TODOs that should be resolved or documented. Nice work!

File: src/adapter/types.ts (hunk 1 above)
Comment (alejoberardino) [reply src/adapter/types.ts]: Completed

File: src/config/index.ts (hunk 2 above)
Comment (alejoberardino) [reply src/config/index.ts]: No, I had already renamed 😄

File: src/config/index.ts (hunk 3 above)
Comment (alejoberardino) [reply src/config/index.ts]: Completed

File: src/config/index.ts (hunk 4 above)
Comment (alejoberardino) [reply src/config/index.ts:594]: Not me I think, I moved this from somewhere else in the code

**Review by austinborn — APPROVED** [review]

**Pull request merged**